ui/src/app/pages/analysis/runtime-panel.tsx

```diff
@@ -1112,11 +1180,12 @@ const RuntimePanel = fp.flow(
     // Use gceConfig when (PD feature is not enabled) OR (increase PD size or update machineType of an existing runtime with PD attached)
     // OR (update an existing runtime with no PD attached).
     // post launch PD when all existing running Runtime shutdown.
-    if (!runtimeCtx.enablePD || (gceExists && !pdSizeReduced)) {
+    if (!runtimeCtx.enablePD || (gceExists && !pdSizeReduced) && gpuConfigDiffType === RuntimeDiffState.NO_CHANGE) {
```

Comment:
I don't see why we need to care specifically about GPUs here - there is no additional state tracking necessary, so it should just follow from the normal diffing process

Reply:
The reason is the PATCH endpoint isn't supported today for updating GPU configs in Leo. To change/remove GPU configs we need to delete and recreate the runtime. Thus, we need to jump to the branch providing gceWithPdConfig.

Reply:
That's also true for a lot of runtime settings, such as `ComputeType`, decreasing disk etc; you don't see those enumerated here. This is because they are already handled via `RuntimeDiffState.NEEDS_DELETE_RUNTIME`

Reply:
Hi, CH. The `gpuConfigDiffType === RuntimeDiffState.NO_CHANGE` predicate doesn't affect the covered cases under RuntimeDiffState.NEEDS_DELETE_RUNTIME here. It affects the `RuntimeDiffState.CAN_UPDATE_WITH_REBOOT `cases. 

For example, if we have an existing GCE runtime with PD. Then we want to increase pd size and change GPU config at same time.  Without the `gpuConfigDiffType === RuntimeDiffState.NO_CHANGE` predicate, It will goto the first branch and use `gceConfig` to call Leo's `updateRuntime()`. But Leo's `updateRuntime()` doesn't support GPU update yet, so the GPU won't be updated. To solve this, we need to let it goto the 2nd branch to use `gceWithPdConfig` and delete the runtime then call Leo's `createRuntime()`.

Reply:
OK, I think I finally understand what's happening here. Basically you're enumerating every situation in which a standard VM requires a delete/create.

Here is what I would suggest instead to make this more understandable. Let me know if this covers all the cases you were considering here:

```
const runtimeNeedsDelete = runtimeDiffs.includes(RuntimeDiffState.NEEDS_DELETE_RUNTIME) ||  runtimeDiffs.includes(RuntimeDiffState.NEEDS_DELETE_PD);
...

if (runtimeCtx.enablePD && (runtimeCtx.pdExists || runtimeNeedsDelete)) {
  return {gceWithPdConfig: {...}}
} else {
  ...
}
```

Reply:
Thanks CH! I think that's much more understandable now! 

Just a minor change in the predicate:
```
if (runtimeCtx.enablePD && (!runtimeCtx.gceExists || runtimeNeedsDelete)) {  
return {gceWithPdConfig: {...}}
} else {
  ...
}
```
Since we should create gce with PD by gceWithPdConfig in an empty env after we enable PD.